src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -20,140 +24,352 @@ import fr.acinq.lightning.utils.sat
+                            leaseFeeBase = LightningCodecs.tu32(input).sat
+                        )
+                    )
+                }
+            }
+        }
+
+        fun write(out: Output) = when (this) {
+            is Basic -> {
+                val encoded = encode()
+                LightningCodecs.writeBigSize(0, out) // tag
+                LightningCodecs.writeBigSize(encoded.size.toLong(), out) // length
+                LightningCodecs.writeBytes(encoded, out)
+            }
+        }
+
+        companion object {
+            fun read(input: Input): FundingLease? {
+                val tag = LightningCodecs.bigSize(input)
+                // We always read the lease data, even if we don't support it: this lets us skip unknown lease types.
+                val content = LightningCodecs.bytes(input, LightningCodecs.bigSize(input))
+                return when (tag) {
+                    0L -> Basic.decode(content)
+                    else -> null
+                }
+            }
+        }
+    }
+
+    sealed class FundingLeaseWitness {
+        abstract val fundingScript: ByteVector
+
+        /** The seller signs a witness that commits to the funding script. */
+        data class Basic(override val fundingScript: ByteVector) : FundingLeaseWitness()
+
+        fun signedData(): ByteArray {
+            val tag = when (this) {
+                is Basic -> "basic_funding_lease"
+            }
+            val tmp = ByteArrayOutput()
+            write(tmp)
+            return Crypto.sha256(tag.encodeToByteArray() + tmp.toByteArray())
+        }
+
+        fun write(out: Output) = when (this) {
+            is Basic -> {
+                LightningCodecs.writeBigSize(0, out) // tag
+                LightningCodecs.writeBigSize(fundingScript.size().toLong(), out) // length
+                LightningCodecs.writeBytes(fundingScript, out)
+            }
+        }
+
+        companion object {
+            fun read(input: Input): FundingLeaseWitness = when (val tag = LightningCodecs.bigSize(input)) {
+                0L -> Basic(LightningCodecs.bytes(input, LightningCodecs.bigSize(input)).byteVector())
+                else -> throw IllegalArgumentException("unknown funding lease witness (tag=$tag)")
+            }
+        }
+    }
+
+    /** The fees associated with a given [LeaseRate] can be paid using various options. */
+    sealed class PaymentType {
+        /** Fees are transferred from the buyer's channel balance to the seller's during the interactive-tx construction. */
+        data object FromChannelBalance : PaymentType()
+        /** Fees will be deducted from future HTLCs that will be relayed to the buyer. */
+        data object FromFutureHtlc : PaymentType()
+        /** Fees will be deducted from future HTLCs that will be relayed to the buyer, but the preimage is revealed immediately. */
+        data object FromFutureHtlcWithPreimage : PaymentType()
```

Comment:
I would completely drop that case.

Reply:
I kept it for now since it's very few additional code, and people wanted to have both options in the spec. We can disable it by default though.